In [1]:
# Data Loading
import pandas as pd
import dataInspectionTools as ditools

df = pd.read_csv('loan_data_2007_2014.csv')
#removing index column
df.drop(['Unnamed: 0'], axis=1, inplace=True)

/var/folders/42/3d9r3rcx2vgd6bmrbmhv51k40000gn/T/ipykernel_66977/1373005959.py:5: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('loan_data_2007_2014.csv')


In [2]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

   int_rate  installment grade sub_grade  ... total_bal_il il_util  \
0     10.65       162.87     B        B2  ...          NaN     NaN   
1     15.27        59.83     C        C4  ...          NaN     NaN   
2     15.96        84.33     C        C5  ...          NaN     NaN   
3     13.49       339.31     C        C1  ...          NaN     NaN   
4     12.69        67.79     B        B5  ...          NaN     NaN   

  open_rv_12m  open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi  \
0         NaN          NaN        NaN      NaN              NaN    NaN   
1         NaN          NaN        NaN      NaN              NaN    NaN   
2         NaN          NaN        NaN      NaN              NaN    NaN   
3         NaN          NaN        NaN      NaN              NaN    NaN   
4         NaN          NaN        NaN      NaN              NaN    NaN   

  total_cu_tl inq_last_12m  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN          NaN  
4         NaN          NaN  

[5 rows x 74 columns]

# Column Category

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 74 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           466285 non-null  int64  
 1   member_id                    466285 non-null  int64  
 2   loan_amnt                    466285 non-null  int64  
 3   funded_amnt                  466285 non-null  int64  
 4   funded_amnt_inv              466285 non-null  float64
 5   term                         466285 non-null  object 
 6   int_rate                     466285 non-null  float64
 7   installment                  466285 non-null  float64
 8   grade                        466285 non-null  object 
 9   sub_grade                    466285 non-null  object 
 10  emp_title                    438697 non-null  object 
 11  emp_length                   445277 non-null  object 
 12  home_ownership               466285 non-null  object 
 13 

## Numerical and categorical data
This step will helps to divide type of data for every columns

In [4]:
#numeric data
numCols, catCols = ditools.getNumericalCategorical(df)
print('there are',str(len(numCols)),'columns of numerical data. including:',str(numCols[:5]))
print('there are',str(len(catCols)),'columns of categorical data. including:',str(catCols[:5]))


there are 52 columns of numerical data. including: ['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv']
there are 22 columns of categorical data. including: ['last_pymnt_d', 'title', 'emp_length', 'zip_code', 'pymnt_plan']


### Notes

<b>52 columns</b> are having numerical data

<b>22 columns</b> are having categorical data


<br><br>
<div class="alert alert-block alert-success">
    <b> The data can be further divided into before and after loan application </b>
    <br>This is important to determine which columns will be used on credit risk assessment
</div>

# Missing Values

This step will help determining which columns are actively being used and the ones that never be used.

## Missing Values Function

## Missing Data Understanding

In [5]:
missings_cal, missing_stat = ditools.missing_info(df)
missings_cal

column_name  missing_count  percent_missing
0                 id              0         0.000000
1          member_id              0         0.000000
2          loan_amnt              0         0.000000
3        funded_amnt              0         0.000000
4    funded_amnt_inv              0         0.000000
..               ...            ...              ...
69          all_util         466285       100.000000
70  total_rev_hi_lim          70276        15.071469
71            inq_fi         466285       100.000000
72       total_cu_tl         466285       100.000000
73      inq_last_12m         466285       100.000000

[74 rows x 3 columns]

In [6]:
missing_stat

missing_category  table_number  \
0          No Null            34   
1        Semi Null            23   
2       Fully Null            17   

                                          table_list  
0  [id, member_id, loan_amnt, funded_amnt, funded...  
1  [emp_title, emp_length, annual_inc, desc, titl...  
2  [annual_inc_joint, dti_joint, verification_sta...

In [7]:
#showing only missing columns
missings_cal[missings_cal['missing_count'] != 0]

column_name  missing_count  percent_missing
10                    emp_title          27588         5.916553
11                   emp_length          21008         4.505399
13                   annual_inc              4         0.000858
19                         desc         340302        72.981546
21                        title             20         0.004289
25                  delinq_2yrs             29         0.006219
26             earliest_cr_line             29         0.006219
27               inq_last_6mths             29         0.006219
28       mths_since_last_delinq         250351        53.690554
29       mths_since_last_record         403647        86.566585
30                     open_acc             29         0.006219
31                      pub_rec             29         0.006219
33                   revol_util            340         0.072917
34                    total_acc             29         0.006219
45                 last_pymnt_d            376         0.080637
47                 next_pymnt_d         227214        48.728567
48           last_credit_pull_d             42         0.009007
49   collections_12_mths_ex_med            145         0.031097
50  mths_since_last_major_derog         367311        78.773926
53             annual_inc_joint         466285       100.000000
54                    dti_joint         466285       100.000000
55    verification_status_joint         466285       100.000000
56               acc_now_delinq             29         0.006219
57                 tot_coll_amt          70276        15.071469
58                  tot_cur_bal          70276        15.071469
59                  open_acc_6m         466285       100.000000
60                   open_il_6m         466285       100.000000
61                  open_il_12m         466285       100.000000
62                  open_il_24m         466285       100.000000
63           mths_since_rcnt_il         466285       100.000000
64                 total_bal_il         466285       100.000000
65                      il_util         466285       100.000000
66                  open_rv_12m         466285       100.000000
67                  open_rv_24m         466285       100.000000
68                   max_bal_bc         466285       100.000000
69                     all_util         466285       100.000000
70             total_rev_hi_lim          70276        15.071469
71                       inq_fi         466285       100.000000
72                  total_cu_tl         466285       100.000000
73                 inq_last_12m         466285       100.000000

In [8]:
#showing only missing columns and not entirely missing (100%)
missings_cal[((missings_cal['missing_count'] != 0) & (missings_cal['percent_missing'] != 100))].sort_values(by='percent_missing', ascending=False)

column_name  missing_count  percent_missing
29       mths_since_last_record         403647        86.566585
50  mths_since_last_major_derog         367311        78.773926
19                         desc         340302        72.981546
28       mths_since_last_delinq         250351        53.690554
47                 next_pymnt_d         227214        48.728567
70             total_rev_hi_lim          70276        15.071469
58                  tot_cur_bal          70276        15.071469
57                 tot_coll_amt          70276        15.071469
10                    emp_title          27588         5.916553
11                   emp_length          21008         4.505399
45                 last_pymnt_d            376         0.080637
33                   revol_util            340         0.072917
49   collections_12_mths_ex_med            145         0.031097
48           last_credit_pull_d             42         0.009007
27               inq_last_6mths             29         0.006219
30                     open_acc             29         0.006219
34                    total_acc             29         0.006219
26             earliest_cr_line             29         0.006219
25                  delinq_2yrs             29         0.006219
56               acc_now_delinq             29         0.006219
31                      pub_rec             29         0.006219
21                        title             20         0.004289
13                   annual_inc              4         0.000858

### TOP 5 missing columns

| Table Name | Table Description | Missing percentage |
| --- | --- | --- |
| mths_since_last_record | The number of months since the last public record. | 84.6% |
| mths_since_last_major_derog | Months since most recent 90-day or worse rating. | 78.8% |
| desc | Loan description provided by the borrower. | 73% |
| mths_since_last_delinq | The number of months since the borrower's last delinquency. | 53.7% |
| next_pymnt_d | Next scheduled payment date | 48.7% |


<b> Hypothesis based on description </b>


- <b>mths_since_last_record</b> \
No activity/input of public record for the user 


- <b>mths_since_last_major_derog</b>\
No input/record for derogatory of the user


- <b>desc</b>\
User give no description for their loans


- <b>mths_since_last_delinq</b>\
No input/record for delinquency of the user


- <b>next_pymnt_d</b>\
No more payments needed (Check 'loan_status' column)

# Business Understanding
This part is to explore the business flow based on the available data

First step, We want to make sure that we only inspect active columns.

this means all the fully null will be eliminated.

In [9]:
fullColumn = missing_stat[missing_stat['missing_category'] == 'No Null'].table_list.values[0]
semiNullColumn = missing_stat[missing_stat['missing_category'] == 'Semi Null'].table_list.values[0]

Next step is to inspect the columns. This might need to be done manually as there is no automated way available yet.

Since the data is regards to loan, we will build a hypothesis first with general knowledge of loans application.

Assuming the model of loan application used by the company is the common way there will be several steps for activity within the business:
1. The user/borrower apply to a loan
2. Additional data regarding user will be acquired
3. Lender will assess the data and gives loan based on the assesment
4. Borrower will repay periodically
5. In the case of borrower are not able to pay in certain amount of time, late fees will be charged
6. Recovery action will be proceeded for defaulted loan

<br><br>
<b> Hypothesis </b>
> The process of data acquiring can be divided into 3 phase: Before loan, while loan active, after loan

This helps to determine which data can be used for early credit risk assesment as we want to focus on what data can be gathered to assess future borrower.

This categorization can be narrowed down into 2 category as we want to focus on assessing the user. These catefories are : Assesment data and loan activity data.

- Assesment data
<br>    this data is acquired <b>before or during</b> the approval of loan
- Loan activity data
<br>    this data is acquired <b>after</b> the approval of loan

## Columns manual inspection
in this phase, the columns will be observed and categorized based on the data acquiring. The decision will be based on the description of the columns.

To improve the clarity, the columns that will be inspected is going to be 5 at a time.

In [10]:
assementDataColumns = []
loanActivityDataColumns = []

### fullColumn category

In [11]:
# 1-5
fullColumn[:5]

['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv']

| Table Name | Table Description | Category |
| --- | --- | --- |
| id | A unique LC assigned ID for the loan listing. | Assesment data |
| member_id | A unique LC assigned Id for the borrower member. | Assesment data |
| loan_amnt | The listed amount of the loan applied for by the borrower. If at some point in time, the credit department reduces the loan amount, then it will be reflected in this value. | Assesment data |
| funded_amnt | The total amount committed to that loan at that point in time. | Loan activity data |
| funded_amnt_inv | (Assumption) The same as funded_amnt but by investor  | Loan activity data |

<br><br>
<b>Notes</b> :
- id and member_id will be assigned directly in case of loan applied and member registered
- loan_amnt might be changed by LC but essentially it filled before the loan is active
- funded_amnt and funded_amnt_inv is the number commited to the loan therefore it's more related to Loan activity data

In [12]:
assementDataColumns.extend(['id', 'member_id', 'loan_amnt'])
loanActivityDataColumns.extend(['funded_amnt', 'funded_amnt_inv'])

In [13]:
# 6-10
fullColumn[5:10]

['term', 'int_rate', 'installment', 'grade', 'sub_grade']

| Table Name | Table Description | Category |
| --- | --- | --- |
| term | The number of payments on the loan. Values are in months and can be either 36 or 60. | Assesment data |
| int_rate | interest rate | Loan activity data |
| installment | The monthly payment owed by the borrower if the loan originates. | Loan activity data |
| grade | LC assigned loan grade | Loan activity data |
| sub_grade | LC assigned loan subgrade | Loan activity data |

<br><br>
<b>Notes</b> :
- since term is more likely is input by the user it categorized to Assesment data
- int_rate and installment indicates of the borrower repay activities therefore it will be Loan activity data
- grade and sub_grade is assigned by LC on the assessment thus belong to Loan activity data.

In [14]:
assementDataColumns.extend(['term'])
loanActivityDataColumns.extend(['int_rate', 'installment', 'grade', 'sub_grade'])

In [15]:
# 6-10
fullColumn[10:15]

['home_ownership',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan']

| Table Name | Table Description | Category |
| --- | --- | --- |
| home_ownership | The home ownership status provided by the borrower during registration. Our values are: RENT, OWN, MORTGAGE, OTHER. | Assesment data |
| verification_status |  |  |
| issue_d | The month which the loan was funded. | Loan activity data |
| loan_status | Current status of the loan | Loan activity data |
| pymnt_plan |  |  |

<br><br>
<b>Notes</b> :
- As home_ownership provided by the user, it belongs to Assesment data
- issue_d and loan_status will be available after the loan approved so it will belong to Loan activity data.
- loan_status column may helps to determine the risk assesment for the borrower
- verification_status and pymnt_plan does not have any description futher inspection needed

In [16]:
df['verification_status'].value_counts()

Verified           168055
Source Verified    149993
Not Verified       148237
Name: verification_status, dtype: int64

In [17]:
df['pymnt_plan'].value_counts()

n    466276
y         9
Name: pymnt_plan, dtype: int64

<b>Further notes</b> :
- verification_status may regards to the confirmation of specific data verification, this still ambiguous, but it can be assumed that it belongs to Assesment data as it will be used for the application approval
- y and n answer only indicates that as a yes and no question, therefore it possibly comes from option that user choose so it can also be assumed that it belongs to Assesment data

In [18]:
assementDataColumns.extend(['home_ownership', 'verification_status', 'pymnt_plan'])
loanActivityDataColumns.extend(['issue_d', 'loan_status'])

In [19]:
# 16-20
fullColumn[15:20]

['url', 'purpose', 'zip_code', 'addr_state', 'dti']

| Table Name | Table Description | Category |
| --- | --- | --- |
| url | URL for the LC page with listing data. | Assesment data |
| purpose | A category provided by the borrower for the loan request.  | Assesment data |
| zip_code | The first 3 numbers of the zip code provided by the borrower in the loan application. | Assesment data |
| addr_state | The state provided by the borrower in the loan application | Assesment data |
| dti | A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income. | Assesment data |

<br><br>
<b>Notes</b> :
- All these 5 are categorized to assesment data, reasoning:
    - url are automatically inputed after the user create a loan application
    - purpose, zip_code, and addr_state are data input provided by the user
    - dti is the calculation using formula at the description

In [20]:
assementDataColumns.extend(fullColumn[15:20])

In [21]:
# 21-25
fullColumn[20:25]

['revol_bal',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt']

| Table Name | Table Description | Category |
| --- | --- | --- |
| revol_bal | Total credit revolving balance. | Loan activity data |
| initial_list_status | The initial listing status of the loan. Possible values are – Whole, Fractional  | Loan activity data |
| out_prncp | Remaining outstanding principal for total amount funded | Loan activity data |
| out_prncp_inv | Remaining outstanding principal for portion of total amount funded by investors | Loan activity data |
| total_pymnt | Payments received to date for total amount funded | Loan activity data |

<br><br>
<b>Notes</b> :
- All these 5 are categorized to Loan activity data, reasoning:
    - revol_bal, out_prncp, out_prncp_inv, and total_pymnt are current status of the loan (after approval)
    - initial_list_status is assumed to be declared after the loan approval

In [22]:
loanActivityDataColumns.extend(fullColumn[20:25])

In [23]:
# 25-30
fullColumn[25:30]

['total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries']

| Table Name | Table Description | Category |
| --- | --- | --- |
| total_pymnt_inv | Payments received to date for portion of total amount funded by investors | Loan activity data |
| total_rec_prncp | Principal received to date | Loan activity data |
| total_rec_int | Interest received to date | Loan activity data |
| total_rec_late_fee | Late fees received to date | Loan activity data |
| recoveries | Indicates if a payment plan has been put in place for the loan | Loan activity data |

<br><br>
<b>Notes</b> :
- 4 columsn are categorized to Loan activity data, reasoning:
    - total_pymnt_inv is similar to total_pymnt which part of it but dedicated for investor's fund
    - total_rec_prncp, total_rec_int, and total_rec_late_fee are dynamic data that updated in every transaction (repay) after loan approval
- recoveries needs to be inspected further

In [24]:
df['recoveries'].value_counts()

0.00       441914
14.28           9
11.70           9
200.00          9
10.00           9
            ...  
423.23          1
2311.77         1
850.18          1
5052.98         1
1159.40         1
Name: recoveries, Length: 22773, dtype: int64

<b>Further notes</b> :
- it seems recoveries description does not match with what data provided but as recovery actions are only taken after loan are defaulted, it safe to assume that it categorized into Loan activity data

In [25]:
loanActivityDataColumns.extend(fullColumn[25:30])

In [26]:
# 31-34
fullColumn[30:]

['collection_recovery_fee',
 'last_pymnt_amnt',
 'policy_code',
 'application_type']

| Table Name | Table Description | Category |
| --- | --- | --- |
| collection_recovery_fee | post charge off collection fee | Loan activity data |
| last_pymnt_amnt | Last total payment amount received | Loan activity data |
| policy_code | publicly available policy_code=1, new products not publicly available policy_code=2 |  |
| application_type | Indicates whether the loan is an individual application or a joint application with two co-borrowers | Assesment data |

<br><br>
<b>Notes</b> :
- collection_recovery_fee and last_pymnt_amnt indicated to be the data that will only have input after the loan approved therefore it will belong into Loan activity data
- policy_code here still does not described clearly as what is piblicly available and what is not. It could be the product type or the option input by user or even the data of user available to be accessed by public. Further exploration might be needed
- application_type indicates what type the loan is therefore the data might available before the approval which belong to assesment data

In [27]:
df['policy_code'].value_counts()

1    466285
Name: policy_code, dtype: int64

<b>Further notes</b> :
- policy_code in this data only consist of 1 value therefore it can be ignored and categorized into inactive data same as columns with missing data

In [28]:
assementDataColumns.extend(['application_type'])
loanActivityDataColumns.extend(['collection_recovery_fee', 'last_pymnt_amnt'])

As a checkpoint we need to make sure all columns from No Null category is already categorized into assesment and loan activity data.

In [29]:
print('assementDataColumns include',str(len(assementDataColumns)),'columns')
print('loanActivityDataColumns include',str(len(loanActivityDataColumns)),'columns')
print('totally all categorized columns are',str(len(loanActivityDataColumns) +len(assementDataColumns)))

assementDataColumns include 13 columns
loanActivityDataColumns include 20 columns
totally all categorized columns are 33


so from 34 columns, 33 already categorized. the 1 column that are not included are the policy_code as it only contain 1 kind of input which will be regarded as inactive column.

### Semi null columns

In [30]:
semiNullColumn[:5]

['emp_title', 'emp_length', 'annual_inc', 'desc', 'title']

| Table Name | Table Description | Category |
| --- | --- | --- |
| emp_title | The job title supplied by the Borrower when applying for the loan. Employer Title replaces Employer Name for all loans listed after 9/23/2013 | Assesment data |
| emp_length | Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.  | Assesment data |
| annual_inc | The self-reported annual income provided by the borrower during registration. | Assesment data |
| desc | Loan description provided by the borrower | Assesment data |
| title | The loan title provided by the borrower | Assesment data |

<br><br>
<b>Notes</b> :
- based on the description, as these 5 columns are input by the user, they will belong to Assesment data

In [31]:
assementDataColumns.extend(semiNullColumn[:5])

In [32]:
semiNullColumn[5:10]

['delinq_2yrs',
 'earliest_cr_line',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record']

| Table Name | Table Description | Category |
| --- | --- | --- |
| delinq_2yrs | The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years | Assesment data |
| earliest_cr_line | The month the borrower's earliest reported credit line was opened  | Assesment data |
| inq_last_6mths | Number of credit inquiries in past 12 months | Assesment data |
| mths_since_last_delinq | The number of months since the borrower's last delinquency | Assesment data |
| mths_since_last_record | The number of months since the last public record | Assesment data |

<br><br>
<b>Notes</b> :
- based on the description, these 5 columns are datas available to be accessed to helps assesing past history of the user hence they will belong to Assesment data

In [33]:
assementDataColumns.extend(semiNullColumn[5:10])

In [34]:
semiNullColumn[10:15]

['open_acc', 'pub_rec', 'revol_util', 'total_acc', 'last_pymnt_d']

| Table Name | Table Description | Category |
| --- | --- | --- |
| open_acc | The number of open credit lines in the borrower's credit file | Assesment data |
| pub_rec | Number of derogatory public records  | Assesment data |
| revol_util | Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit. | Loan activity data |
| total_acc | The total number of credit lines currently in the borrower's credit file | Loan activity data |
| last_pymnt_d | Last month payment was received | Loan activity data |

<br><br>
<b>Notes</b> :
- open_acc and pub_rec are the data in which will be available to be accessed pre-approval of the loan therefore it categorized to Assesment data
- revol_util, total_acc, and last_pymnt_d indicated to be current updated data of the borrower's therefore it belong into Loan activity data.

In [35]:
assementDataColumns.extend(['open_acc', 'open_acc'])
loanActivityDataColumns.extend(['revol_util', 'total_acc', 'last_pymnt_d'])

In [36]:
semiNullColumn[15:20]

['next_pymnt_d',
 'last_credit_pull_d',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'acc_now_delinq']

| Table Name | Table Description | Category |
| --- | --- | --- |
| next_pymnt_d | Next scheduled payment date | Loan activity data |
| last_credit_pull_d |   |  |
| collections_12_mths_ex_med | Number of collections in 12 months excluding medical collections | Loan activity data |
| mths_since_last_major_derog | Months since most recent 90-day or worse rating | Assesment data |
| acc_now_delinq | The number of accounts on which the borrower is now delinquent. | Assesment data |

<br><br>
<b>Notes</b> :
- next_pymnt_d will be available if the user is a borrower, therefore it belongs to Loan activity data
- last_credit_pull_d does not have a description, further exploration needed
- collections_12_mths_ex_med can be assumed as the collection belong to current loan therefore it will be a Loan activity data
- mths_since_last_major_derog and acc_now_delinq, assuming these columns are notes or record available to be accessed it wull belong to Assesment data

In [37]:
df['last_credit_pull_d'].value_counts

<bound method IndexOpsMixin.value_counts of 0         Jan-16
1         Sep-13
2         Jan-16
3         Jan-15
4         Jan-16
           ...  
466280    Jan-16
466281    Jan-16
466282    Dec-15
466283    Apr-15
466284    Jan-16
Name: last_credit_pull_d, Length: 466285, dtype: object>

In [38]:
df[['last_pymnt_d','last_credit_pull_d']][:10]

last_pymnt_d last_credit_pull_d
0       Jan-15             Jan-16
1       Apr-13             Sep-13
2       Jun-14             Jan-16
3       Jan-15             Jan-15
4       Jan-16             Jan-16
5       Jan-15             Sep-15
6       Jan-16             Jan-16
7       Jan-15             Dec-14
8       Apr-12             Aug-12
9       Nov-12             Mar-13

Assuming the data format is mmm-yy, it seems last_credit_pull_d is the activity while loan is active, therefore it will belong inti Loan activity data

In [39]:
assementDataColumns.extend(['next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med'])
loanActivityDataColumns.extend(['mths_since_last_major_derog', 'acc_now_delinq'])

In [40]:
semiNullColumn[20:]

['tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim']

| Table Name | Table Description | Category |
| --- | --- | --- |
| tot_coll_amt |  | Loan activity data |
| tot_cur_bal |   | Loan activity data |
| total_rev_hi_lim | Total revolving high credit/credit limit | Loan activity data |

<br><br>
<b>Notes</b> :
- based on the name structure, these three columns can be assumed to be a recorded data while the loan is active. therefore they belong to Loan activity data

In [41]:
loanActivityDataColumns.extend(semiNullColumn[20:])

As another checkpoint, we will make sure that all category does not have any duplicate.

In [42]:
print('assementDataColumns include',str(len(assementDataColumns)),'columns')
print('loanActivityDataColumns include',str(len(loanActivityDataColumns)),'columns')
print('totally all categorized columns are',str(len(loanActivityDataColumns) +len(assementDataColumns)))

assementDataColumns include 28 columns
loanActivityDataColumns include 28 columns
totally all categorized columns are 56


With 34 full columns and 23 columns semi null, there will be total of 57 columns. but since the policy_code column disregarded, there will be total of 56 columns

## Saving column category
to enable the easier access to the list we will create a DataFrame then save it into a csv file

In [43]:
dict = {'Category': ['Assesment Data', 'Loan Activity data'], 'columns': [assementDataColumns, loanActivityDataColumns]}  

columnCatDF = pd.DataFrame(dict) 
columnCatDF

Category                                            columns
0      Assesment Data  [id, member_id, loan_amnt, term, home_ownershi...
1  Loan Activity data  [funded_amnt, funded_amnt_inv, int_rate, insta...

In [44]:
columnCatDF.to_csv('columnCategory.csv', index=False) 